[로또] 이바 미션 제출합니다

seonghun127 opened the pull request.

seonghun127 committed:
docs : 기능 구현 목록 정리

seonghun127 committed:
docs : 기능 구현 목록 정리

seonghun127 committed:
feat : 로또 번호 생성 기능 구현

1. 1-45 숫자를 하나씩 가지는 로또 번호 객체 총 45개를 미리 만들어서
List 에 저장한다
2. 로또 번호 객체를 접근하기 위해 get 메소드를 static으로 구현한다.

seonghun127 committed:
refactor : 로또 번호범위 리펙토링

1. LottoNumber 클래스에서  1이상 45이하의 번호에 대해 제대로 생성하도록 인덱스를 리펙토링한다
	- 로또 번호와 numbers의 인덱스가 일치하지 않기 때문

seonghun127 committed:
feat : 로또 6자리를 담고 관리하는 Lotto 클래스 추가

1. LottoNumber 타입의 리스트를 인자로 받고 6자리가 맞는지 확인한다
	- 6개의 숫자가 아니거나 중복일 경우 예외를 던진다

seonghun127 committed:
feat : 로또 번호 6개 조합을 만드는 기능 추가

1. LottoFactory 클래스에서 로또 번호 6자리 조합을 만든다.
2. 로또 번호는 LottoNumber 클래스에서 Collection.shuffle() 메소드를 통해
무작위 번호를 반환한다

seonghun127 committed:
feat : 로또 당첨 금액을 담고있는 LottoRank enum 추가

1. 번호 일치 개수와 금액을 연관지어 등수에 지정한다

seonghun127 committed:
refactor : 코드 포맷팅

seonghun127 committed:
feat : 당첨 번호 입력, 일치 번호 개수, 등수를 구하는 기능 추가

1. LottoFactory 클래스에서 수동으로 로또 번호를 생성한다
2. Lotto 클래스에서 당첨 로또 번호와 일치하는 개수를 구한다
3. WinningLotto 클래스에서 일치하는 번호 개수에 따른 등수를 반환한다

seonghun127 committed:
refactor : 로또 번호 6자리 저장방식을 Set 구조로 변경

1. 로또 번호 자동 생성시 중복 조합이 발생할 경우를 처리하기 위해
List 자료구조에서 Set으로 변경한다

seonghun127 committed:
refactor : WinningLottoTest 클래스 리펙토링

1. numbers Set 구조 생성의 중복을 없애기 위해 setUp 메소드와 tearDown 메소드를 추가한다

seonghun127 committed:
feat : Lotto 게임 진행을 담당하는 WinningResult와 LottoGame 클래스 추가

1. 당첨 번호와 생성된 로또를 비교하여 Map 구조에 결과를 저장하는 기능을
LottoGame 클래스에 추가한다
2. 로또 결과를 가지고 총 당첨 금액과 수익률를 계산하는 기능을 WinningResult
클래스에 추가한다
3. 위 두개의 클래스를 WinningResultTest 클래스에서 테스트한다

seonghun127 committed:
feat : Input과 Output을 담당하는 View 클래스 추가

seonghun127 committed:
refactor : 로또 한장 발급 시 오름차순으로 번호 정렬

1. HashSet으로 관리하던 로또 번호 6자리를 TreeSet으로 저장하여
오름차순 정렬한다

seonghun127 committed:
refactor : 등수를 순서대로 출력하기 위해 LinkedHashMap 사용

seonghun127 committed:
docs : 기능 구현 목록 갱신

seonghun127 committed:
feat : 2등 보너스볼을 처리하는 기능을 추가

seonghun127 committed:
refactor : 보너스 볼기능 추가에 대한 테스트 코드 리펙토링

seonghun127 committed:
docs : 기능 구현 목록 갱신

seonghun127 committed:
refactor : 수동, 자동으로 생성된 로또를 관리하고 결과를 반환하는 기능 리팩토링

1. Lottos 클래스에서 lotto 리스트를 생성한다
2. 생성된 lotto 리스트에 대한 결과 값을 저장한다

seonghun127 committed:
refactor : 오타 수정

seonghun127 committed:
refactor : Lottos 클래스 추가에 따른 테스트 코드 수정

seonghun127 committed:
feat : Input, Output 기능 추가

seonghun127 committed:
feat : main 담당 클래스 추가

seonghun127 committed:
refactor : 코드 포맷팅

seonghun127 committed:
feat : NumberException 예외 처리 클래스 추가

seonghun127 committed:
refactor : 코드 포맷팅

seonghun127 committed:
refactor : 코드 포맷팅

seonghun127 committed:
refactor : indent 2 이상인 로직을 1로 리펙토링

gracefulBrown commented on src/main/java/lotto/domain/LottoFactory.java:
```diff
@@ -0,0 +1,18 @@
+
+    public static Lotto createLottoManually(String manualNumbers) {
+        Set<LottoNumber> numbers = new TreeSet<>();
+        for (String manualNumber : manualNumbers.split(",")) {
```
manualNumbers.split(",") 내부에서 split을 통해 구분 받는 방법으로 구현 할 경우
LottoFactory가 입력 형식에 종속되는 현상이 발생합니다.
예) 콘솔에서는 1,2,3,4,5,6 으로 받지만 웹에서는 6개의 input 태그를 통해 각각 받는 경우 

입력방법과 무관하게 createLottoManually를 구현해보면 어떨까요?

seonghun127 replied:
위 메소드의 매개변수 타입을 `List<String>`으로 변경해보았습니다. 이를 통해 입력 형식에 종속되는 문제를 해결할 수 있을거라고 생각이 되었는데요. 입력방식이 `String`으로 "1,2,3,4,5,6"으로 들어오나 "1", "2", "3", "4", "5", "6"으로 들어오나 컨트롤러에서 `List<String>` 타입으로 변경해주고 해당 메서드를 호출하면 된다는 생각을 했기때문입니다.
어쨋든 컨트롤러는 어떠한 입력 방식으로 들어오던 해당 입력 방식을 처리하는 메서드를 호출하기 전에 그 메서드의 매개변수의 형식으로 변경을 해야하는 과정이 일어나기 된다고 생각되었습니다.
저의 생각과 리펙토링한 방식이 맞다고 볼 수 있을까요?

gracefulBrown commented on src/main/java/lotto/domain/LottoFactory.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.Set;
+import java.util.TreeSet;
+
+public class LottoFactory {
```
Lotto내부에서 정적 팩토리 메서드를 통해 자동과 수동인 경우를 생성하면 어떨까요?
LottoFactory가 필요 없을 수 있을 것 같아서요..ㅎㅎ

seonghun127 replied:
Lotto 클래스에서 자동과 수동의 경우를 생성하는 말이 잘 이해가 되지 않습니다 ㅠ
자/수동으로 로또를 생성해주는 LottoFactory 클래스의 역할을 Lotto 클래스로 이동시키라는 말씀이신가요?

gracefulBrown commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,53 @@
+public class Lotto {
+    private static final int LOTTO_SIZE = 6;
+
+    private final Set<LottoNumber> numbers;
```
Set을 이용하는군요 👍

gracefulBrown commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,59 @@
+    private static final int LAST_INDEX_AFTER_SHUFFLED = 6;
+
+    private final int number;
+    private static final List<LottoNumber> numbers;
```
여러 번 사용되는 LottoNumber를 static 변수로 만들었네요 !
numbers에 대해서 고민을 많이 한 것 같아요 👍 

numbers가 현재 LottoNumber 내부에 위치하고 있는데 이 점에 대해서는 어떻게 생각하세요?
어쩌면 LottoFactory의 역할이 Lotto를 생성해주는 것이 아니라 LottoNumber를 할당해주는 것 일 수도 있겠네요ㅎㅎ

seonghun127 replied:
원래는 numbers를 클래스로 따로 빼려고했는데 유효성 검사에서 LottoNumber 클래스와 중복되는 부분이 많이 발생하여 LottoNumber 클래스가 numbers List 를 가지고 있어도 되겠다는 생각이 들어 내부에 위치시켰습니다!

gracefulBrown commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,59 @@
+        this.number = number;
+    }
+
+    public static LottoNumber get(int number) {
```
미리 만들어놓은 LottoNumber를 찾는 메서드군요!
혹시 ENUM으로 미리 만들어 놓으면 이런 로직이 필요 없을 수도 있겠네요ㅎㅎ

seonghun127 replied:
1부터 45의 로또 숫자를 가지는 ENUM 말씀이신가요?!

gracefulBrown commented on src/main/java/lotto/domain/LottoRank.java:
```diff
@@ -0,0 +1,45 @@
+        }
+
+        return Arrays.stream(values())
+                .filter(lottoRank -> (lottoRank.countOfMatch == countOfMatch) && lottoRank != SECOND)
```
filter두번을 쓰는게 좋은지 이바처럼 && 을 쓰는게 좋은지 생각하던차에
그냥 메서드로 분리하면 좋겠다는 생각이 들었네요ㅎㅎㅎ
좋은지는 잘 모르겠지만 추후에 참고해보시면 좋을 것 같아요 :)
```
        return Arrays.stream(values())
                .filter(lottoRank -> lottoRank.matchCountExceptSecondWin(countOfMatch))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("유효한 값이 아닙니다."));```

seonghun127 replied:
코치님의 피드백을 보고 예시에 들어주신 `matchCountExceptSecondWin()` 메소드를 
```
if ((SECOND.countOfMatch == countOfMatch) && bonusMatch) {
            return SECOND;
}
```
이곳에도 동시에 적용할 수 없을까 고민하다가 인자로 로또번호 일치 갯수 `countOfMatch` 뿐만 아니라 추가적인 boolean 값 `additionalCheck`을 받아서 두곳 모두에 적용할 수 있는 메소드를 추가해보았습니다.
```
private boolean matchCount(int countOfMatch, boolean additionalCheck) {
        return (this.countOfMatch == countOfMatch && additionalCheck);
}
```

gracefulBrown commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,59 @@
+                        : 0);
+    }
+
+    private void createLottosManually(List<String> manualNumbers) {
```
void 리턴 메서드에서 값을 변경하는 로직이 있는 경우 사이드이펙트가 발생 할 가능성이 생기며 해당 메서드를 테스트하기 어려울 수 있습니다. 리턴을 한 후 호출한 쪽에서 add를 하는 것은 어떨까요?

seonghun127 replied:
코드를 다시 보면서 Lottos 클래스에서 `List<Lottos> lottos `를 생성하고 있다는게 클래스가 서로 다른 두가지 일을 하고 있다고 판단되었습니다. 원래 Lottos 클래스의 `match()` 메소드로 당첨 결과를 계산하는 역할을 수행하려 했었기 때문에 '생성'이라는 기능을 빼게됐습니다. 이를 LottoFactory 클래스로 옮기고 Lottos 생성자에서는 `List<Lotto> lottos` 를 초기화하는 방식으로 리펙토링했습니다.

gracefulBrown commented on src/main/java/lotto/domain/WinningResult.java:
```diff
@@ -0,0 +1,39 @@
+        for (LottoRank lottoRank : result.keySet()) {
+            sb.append(lottoRank.getCountOfMatch());
+            sb.append("개 일치 ");
+            String second = lottoRank.equals(LottoRank.SECOND) ? ", 보너스볼 일치" : "";
```
도메인 로직이 뷰에 종속적이게 되면 새로운 뷰가 생길 때 마다 수정이 불가피 합니다ㅠㅠ

seonghun127 replied:
get 메서드를 사용하지 않고 구현하려고 노력했는데 다른 방법이 생각이 나지 않아 get 메서드로 OutputView에 전달하여 출력 로직을 구현했습니다

gracefulBrown commented on src/main/java/lotto/domain/LottoRank.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Arrays;
+
+public enum LottoRank {
+    FIRST(6, 2_000_000_000),
```
숫자 표현 시 언더바 사용 👍

gracefulBrown commented on src/test/java/lotto/domain/LottoFactoryTest.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Test
+    public void 로또_6자리_번호_리스트_수동_생성_테스트() {
```
중복 번호 수동 생성 테스트도 있으면 좋을 것 같아요 ㅎㅎ

seonghun127 replied:
Lotto 클래스에서 중복 확인을 하기 때문에 LottoTest에 테스트 코드를 추가했습니다!

gracefulBrown reviewed (changes requested):
안녕하세요. 리뷰어 브라운 입니다.
전체적으로 많은 고민을 한 흔적이 있네요 :)
도메인 분리도 명료하고 잘 하셨습니다 👍 
몇가지 코멘트 남겼는데 확인 부탁드립니다ㅎㅎ
궁금한점 있으면 언제든 DM 주세요 :)

seonghun127 committed:
refactor : LottoFactory 클래스의 createLottoManually 메서드의 매개변수 타입 변경

1. String 타입으로 받던 매개변수 타입을 List<String> 타입으로 변경한다
- 로또 번호가 입력되는 방식에 상관없이 로또를 생성하기 위해

seonghun127 committed:
refactor : 로또 생성 시 중복 숫자에 대한 예외 처리 테스트 코드 추가

seonghun127 committed:
refactor : 문자 상수 사용

seonghun127 committed:
refactor : LottoFactory 클래스에 Lotto 리스트를 생성하는 메서드 추가

1. Lottos 클래스의 생성자에서 생성하던 Lotto 리스트를 LottoFactory의 createLotto 메서드로 생성한다
- Lottos 클래스는 일급 클래스로 당첨 번호를 가지고 당첨 여부를 확인하는 기능( match() ) 외 생성이라는 기능을
가지는 것은 서로 다른 두가지 일을 한다고 판단되어 생성이라는 역할을 LottoFactory 클래스로 위임

2. Lottos 클래스 생성자에서는 이미 생성된 Lotto 리스트를 인자로 받아 인스턴스 필드를 초기화하는 역할을 한다

seonghun127 committed:
refactor : LottoFactory 클래스의 createLottos 메서드 로직 변경

1. 총 구입 횟수와 수동 구입 횟수를 계산하여 자동 로또 생성 횟수를 계산하는 로직을 제거한다

seonghun127 committed:
refactor : delimiter 상수 추가

seonghun127 committed:
refactor : LottoRank enum에 matchCount() 메소드 추가

1. 로또 번호 일치 갯수와 그외 추가로 확인해야하는 boolean 값을 인자로 받는 메서드를 추가한다
2. 추가로 확인하는 boolean 값에는 bonus 번호 일치 여부, SECOND enum 일치 여부가 있다
- boolean bonusMatch
- boolean lottoRank != SECOND

seonghun127 committed:
refactor : WinningResult 클래스에 get 메서드 추가

1. WinningResult 클래스의 toString 재정의를 도메인과 뷰와의 독립을 위해 제거한다
2. get 메서드로 OutputView에 전달하여 outputWinningResult 메서드에서
 당첨 통계 출력 로직을 구현한다

gracefulBrown commented on src/main/java/lotto/domain/WinningResult.java:
```diff
@@ -0,0 +1,31 @@
+        return totalWinningAmount;
+    }
+
+    public Map<LottoRank, Integer> getResult() {
```
👍

gracefulBrown commented on src/main/java/lotto/domain/WinningResult.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    public double calculateRevenueRate(int purchaseAmount) {
+        return (sumTotalWinningAmount() / purchaseAmount) * HUNDRED;
```
도메인에 메서드를 만들어서 뷰에서 사용하는 것도 좋지만 가급적 뷰에 종속적이지 않게 하기위해서는
도메인의 변수값을 getter 통해 뷰에서 사용하는 방향으로 고민해보면 좋겠네요 :)

seonghun127 replied:
알겠습니다! 앞으로 뷰와 도메인의 사이의 독립성을 위해서는 도메인 변수값에 대해서 getter를 사용해야겠네요!

gracefulBrown reviewed (approved):
피드백 반영 잘 해주셨어요 👍 
혹시 피드백 관련해서 의견이나 질문이 있으면
피드백에 코멘트를 바로 남겨줘도 좋고 DM을 줘도 좋아요 :)
수고하셨습니다~
